unit-tests/src/test/scala/java/util/FormatterUSSuite.scala

```diff
@@ -0,0 +1,3535 @@
+        f.format("%-1$s", "1", "2"))
+    }
+
+    locally {
+      val f = new Formatter(Locale.US)
+      assertThrows[UnknownFormatConversionException](
+        f.format("%$s", "hello", "2"))
+    }
+
+    locally {
+      val f = new Formatter(Locale.US)
+      assertThrows[UnknownFormatConversionException](f.format("%", "string"))
+    }
+
+    locally {
+      val f = new Formatter(Locale.US)
+      f.format("%1$s%2$s%3$s%4$s%5$s%6$s%7$s%8$s%<s%s%s%<s",
+               "1",
+               "2",
+               "3",
+               "4",
+               "5",
+               "6",
+               "7",
+               "8",
+               "9",
+               "10",
+               "11")
+      assertEquals("123456788122", f.toString())
+    }
+
+    locally {
+      val f = new Formatter(Locale.US)
+      f.format("xx%1$s22%2$s%s%<s%5$s%<s&%7$h%2$s%8$s%<s%s%s%<ssuffix",
+               "1",
+               "2",
+               "3",
+               "4",
+               "5",
+               "6",
+               7.asInstanceOf[Object], // this is intended
+               "8",
+               "9",
+               "10",
+               "11")
+      assertEquals("xx12221155&7288233suffix", f.toString())
+      assertThrows[MissingFormatArgumentException](f.format("%<s", "hello"))
+    }
+
+    locally {
+      val f = new Formatter(Locale.US)
+      assertThrows[MissingFormatArgumentException](f.format("%123$s", "hello"))
+    }
+
+    locally {
+      val f = new Formatter(Locale.US)
+      // 2147483648 is the value of Integer.MAX_VALUE + 1
+      assertThrows[MissingFormatArgumentException](
+        f.format("%2147483648$s", "hello"))
+      // 2147483647 is the value of Integer.MAX_VALUE
+      assertThrows[MissingFormatArgumentException](
+        f.format("%2147483647$s", "hello"))
+    }
+
+    locally {
+      val f = new Formatter(Locale.US)
+      assertThrows[MissingFormatArgumentException](f.format("%s%s", "hello"))
+    }
+
+    locally {
+      val f = new Formatter(Locale.US)
+      f.format("$100", 100.asInstanceOf[Object])
+      assertEquals("$100", f.toString())
+    }
+
+    locally {
+      val f = new Formatter(Locale.US)
+      f.format("%01$s", "string")
+      assertEquals("string", f.toString())
+    }
+  }
+
+  test("format(String, Array[Object]) for width") {
+    locally {
+      val f = new Formatter(Locale.US)
+      f.format("%1$8s", "1")
+      assertEquals("       1", f.toString())
+    }
+
+    locally {
+      val f = new Formatter(Locale.US)
+      f.format("%1$-1%", "string")
+      assertEquals("%", f.toString())
+    }
+
+    locally {
+      val f = new Formatter(Locale.US)
+      // 2147483648 is the value of Integer.MAX_VALUE + 1
+      f.format("%2147483648s", "string")
+      assertEquals("string", f.toString())
+    }
+
+    // the value of Integer.MAX_VALUE will allocate about 4G bytes of
+    // memory.
+    // It may cause OutOfMemoryError, so this value is not tested
```

Comment:
Turned out that the comment is just copy-pasted from Harmony. I haven't tested whether it actually eats that much memory or not.